File: build_index.js
const fs = require('fs');
const path = require('path');

const pagesDir = path.join(__dirname, 'islamhub', 'js', 'data', 'alquran', 'pages');
const outputFile = path.join(__dirname, 'islamhub', 'js', 'data', 'alquran', 'quran-search-index.js');

let searchIndex = [];

for (let i = 1; i <= 604; i++) {
    const pageNum = String(i).padStart(3, '0');
    const filePath = path.join(pagesDir, `Page${pageNum}.json`);
    
    if (fs.existsSync(filePath)) {
        try {
            const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
            if (data.surahs) {
                for (let s = 0; s < data.surahs.length; s++) {
                    const surah = data.surahs[s];
                    if (surah.verses) {
                        for (let v = 0; v < surah.verses.length; v++) {
                            const verse = surah.verses[v];
                            // Array format to save space:
                            // [surah_num, verse_num, arabic, translation]
                            // Also need page number so clicking takes us to the page!
                            searchIndex.push([
                                surah.number,
                                verse.number,
                                verse.translation,
                                verse.arabic,
                                data.pageNumber
                            ]);
                        }
                    }
                }
            }
        } catch (e) {
            console.error(`Error parsing ${filePath}: ${e.message}`);
        }
    }
}

const jsContent = `// Auto-generated search index for IslamHub
window.QURAN_SEARCH_INDEX = ${JSON.stringify(searchIndex)};
`;

fs.writeFileSync(outputFile, jsContent, 'utf8');
console.log(`Generated search index at ${outputFile}. Total verses: ${searchIndex.length}`);

File: build_index.py
import os
import json

base_dir = os.path.dirname(os.path.abspath(__file__))
pages_dir = os.path.join(base_dir, 'islamhub', 'js', 'data', 'alquran', 'pages')
output_file = os.path.join(base_dir, 'islamhub', 'js', 'data', 'alquran', 'quran-search-index.js')

search_index = []

for i in range(1, 605):
    page_num = str(i).zfill(3)
    file_path = os.path.join(pages_dir, f'Page{page_num}.json')
    
    if os.path.exists(file_path):
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                
                if 'surahs' in data:
                    for surah in data['surahs']:
                        if 'verses' in surah:
                            for verse in surah['verses']:
                                search_index.append([
                                    surah['number'],
                                    verse['number'],
                                    verse.get('translation', ''),
                                    verse.get('arabic', ''),
                                    data['pageNumber']
                                ])
        except Exception as e:
            print(f"Error parsing {file_path}: {e}")

js_content = f"// Auto-generated search index for IslamHub\nwindow.QURAN_SEARCH_INDEX = {json.dumps(search_index, ensure_ascii=False)};\n"

with open(output_file, 'w', encoding='utf-8') as f:
    f.write(js_content)

print(f"Generated search index at {output_file}. Total verses: {len(search_index)}")
